src/Workspaces/Core/MSBuild/MSBuild/MSBuildProjectLoader.Worker_ResolveReferences.cs

```diff
@@ -0,0 +1,295 @@
+    {
+        private partial class Worker
+        {
+            private struct ResolvedReferences
+            {
+                public ImmutableArray<ProjectReference> ProjectReferences { get; }
+                public ImmutableArray<MetadataReference> MetadataReferences { get; }
+
+                public ResolvedReferences(ImmutableArray<ProjectReference> projectReferences, ImmutableArray<MetadataReference> metadataReferences)
+                {
+                    ProjectReferences = projectReferences;
+                    MetadataReferences = metadataReferences;
+                }
+            }
+
+            private class MetadataReferenceSet
+            {
+                private readonly ImmutableArray<MetadataReference> _metadataReferences;
+                private readonly ImmutableDictionary<string, HashSet<int>> _pathToIndexMap;
+                private readonly HashSet<int> _indecesToRemove;
+
+                public MetadataReferenceSet(ImmutableArray<MetadataReference> metadataReferences)
+                {
+                    _metadataReferences = metadataReferences;
+                    _pathToIndexMap = CreatePathToIndexMap(metadataReferences);
+                    _indecesToRemove = new HashSet<int>();
+                }
+
+                private static ImmutableDictionary<string, HashSet<int>> CreatePathToIndexMap(IEnumerable<MetadataReference> metadataReferences)
+                {
+                    var pathToIndexMap = ImmutableDictionary.CreateBuilder<string, HashSet<int>>(PathUtilities.Comparer);
+
+                    var index = 0;
+                    foreach (var metadataReference in metadataReferences)
+                    {
+                        var filePath = GetFilePath(metadataReference);
+                        if (!pathToIndexMap.TryGetValue(filePath, out var indeces))
+                        {
+                            indeces = new HashSet<int>();
+                            pathToIndexMap.Add(filePath, indeces);
+                        }
+
+                        indeces.Add(index);
+
+                        index++;
+                    }
+
+                    return pathToIndexMap.ToImmutable();
+                }
+
+                public bool Contains(string filePath) => _pathToIndexMap.ContainsKey(filePath);
+
+                public void Remove(string filePath)
+                {
+                    if (filePath != null && _pathToIndexMap.TryGetValue(filePath, out var indexSet))
+                    {
+                        _indecesToRemove.AddRange(indexSet);
+                    }
+                }
+
+                public ImmutableArray<MetadataReference> GetRemaining()
+                {
+                    var results = ImmutableArray.CreateBuilder<MetadataReference>(initialCapacity: _metadataReferences.Length);
+
+                    var index = 0;
+                    foreach (var metadataReference in _metadataReferences)
```

Comment:
why not just use for?

Reply:
because I foolishly refactored to this code and didn't have a refactoring to convert from foreach->for. :smile:

Will change -- thanks!